### PR TITLE
broadcast callback on non shared variables

### LINKED ISSUE #1207: Possibility to forward a change to a subset of clients (conditional broadcast)
**Description**
Being able to broadcast variables not to all the clients but only a subset. A user variable or filter function defines this subset.

**Acceptance Criteria**
- [ ] Ensure new code is unit tested, and check code coverage is at least 90%
- [ ] Propagate any change on the demos and run all of them to ensure there is no breaking change
- [ ] Ensure any change is well documented

### PR DESCRIPTION
**Pull request by FredLL-Avaiga**
resolves #1207

~renamed current `broadcast_callback` to `broadcast_callback_on_shared`~

### CODE REVIEW

**Comment by FabienLelaquais** [comment]
I think this PR is also trying to fix something that was a glitch.

@dinhlongviolin1 I need your blessing here.
This `broadcast_callback_on_shared()` to me is useless: the 'shared' state is not known by users, and should actually never been exposed.
What I expect `broadcast_callback()` to do is invoke a user-defined function, potentially with parameters, for every live state.
If then the user invokes `state.assign()` fine, so be it. But let the user decide.
What do you think?

Another important question is: does anyone see a use case for the `_on_shared` version?
Because I don't see any, and I'm willing to drop all this code...

Thanks!

**Comment by dinhlongviolin1** [comment]
> I think this PR is also trying to fix something that was a glitch.
> 
> @dinhlongviolin1 I need your blessing here. This `broadcast_callback_on_shared()` to me is useless: the 'shared' state is not known by users, and should actually never been exposed. What I expect `broadcast_callback()` to do is invoke a user-defined function, potentially with parameters, for every live state. If then the user invokes `state.assign()` fine, so be it. But let the user decide. What do you think?
> 
> Another important question is: does anyone see a use case for the `_on_shared` version? Because I don't see any, and I'm willing to drop all this code...
> 
> Thanks!

I think you are right on this one and I agree. We should drop this code as I myself don't see a use case for it. Don't remember why I add it but hopefully it wont affect anything. @FredLL-Avaiga Please help me drop it in this PR 😺

**Comment by FredLL-Avaiga** [comment]
> > I think this PR is also trying to fix something that was a glitch.
> > 
> > @dinhlongviolin1 I need your blessing here. This `broadcast_callback_on_shared()` to me is useless: the 'shared' state is not known by users, and should actually never been exposed. What I expect `broadcast_callback()` to do is invoke a user-defined function, potentially with parameters, for every live state. If then the user invokes `state.assign()` fine, so be it. But let the user decide. What do you think?
> > 
> > Another important question is: does anyone see a use case for the `_on_shared` version? Because I don't see any, and I'm willing to drop all this code...
> > 
> > Thanks!
> 
> I think you are right on this one and I agree. We should drop this code as I myself don't see a use case for it. Don't remember why I add it but hopefully it wont affect anything. @FredLL-Avaiga Please help me drop it in this PR 😺 

I will drop it then

File: doc/gui/examples/broadcast.py
Comment (FabienLelaquais) [inline doc/gui/examples/broadcast.py]:
```suggestion
Gui(page).run()
```

File: doc/gui/examples/broadcast.py
Comment (FabienLelaquais) [inline doc/gui/examples/broadcast.py]: The legacy code (and comment) are appropriate and should not be changed:
we want to demonstrate that a shared variable (defined as such in `Gui.add_shared_variable("button_text")`) can be propagated to all clients in a single (and simple) line of code.

File: doc/gui/examples/broadcast.py
Comment (FabienLelaquais) [inline doc/gui/examples/broadcast.py]:
```suggestion
# Invoked by the timer, for every client
def update_counter(state: State, var_name: str, val):
    # Update the variable value for this client
```

The function name indicates what it does, in the context of the invocation.

File: doc/gui/examples/broadcast.py
Comment (FabienLelaquais) [inline doc/gui/examples/broadcast.py]:
```suggestion
        broadcast_callback(gui, update_counter, ["counter", counter])
```

File: frontend/taipy-gui/src/components/Taipy/PaginatedTable.tsx
Comment (FabienLelaquais) [inline frontend/taipy-gui/src/components/Taipy/PaginatedTable.tsx]: Because this may be read by programmers, is there any "better" information we could provide instead of the cryptic "PTable"?

File: taipy/gui/gui.py
Comment (FabienLelaquais) [inline taipy/gui/gui.py:1541]: 👍

File: doc/gui/examples/broadcast.py
Comment (FabienLelaquais) [inline doc/gui/examples/broadcast.py]:
```suggestion
# using a direct assignment of the state property because the variable is declared 'shared'.
```
Should be the truth.

File: doc/gui/examples/broadcast.py
Comment (FabienLelaquais) [inline doc/gui/examples/broadcast.py]:
```suggestion
# The value is updated for every client in a function invoked by broadcast_callback().
```

**Review by FabienLelaquais — CHANGES_REQUESTED** [review]
I would like to be closer to the legacy example.

**Comment by FredLL-Avaiga** [comment]
can't approve as I am still an assignee but I would.
@FabienLelaquais I thought you wanted to add a `Gui.broadcast(var_name: str, value: t.Any)` that would call `Gui.broadcast_callback` ?

**Comment by FredLL-Avaiga** [comment]
I would if you update the branch ;-)

**Comment by FabienLelaquais** [comment]
> I would if you update the branch ;-)

Let me confess I completely forgot.
Thanks!

File: taipy/gui/gui.py
Comment (FredLL-Avaiga) [inline taipy/gui/gui.py]: I would have used kwargs, no ?
A bit less directive maybe but much more pythonic

File: taipy/gui/gui.py
Comment (FredLL-Avaiga) [inline taipy/gui/gui.py:1578]: should be defined as a static method

**Review by FredLL-Avaiga — COMMENTED** [review]
👍

**Review by FabienLelaquais — DISMISSED** [review]
Looking good!

**Review by FabienLelaquais — DISMISSED** [review]
➕

**Review by namnguyen20999 — DISMISSED** [review]
LGTM

**Review by dinhlongviolin1 — DISMISSED** [review]
Amazingly done!

**Review by FabienLelaquais — APPROVED** [review]
👍

**Pull request merged**
